### ISSUE TITLE
Support for union of constants is not allowed in service resource function query and header parameters

### ISSUE THREAD

**Issue by aneeshafedo** [ballerina-platform/ballerina-standard-library#4374]
**Description:**

The singleton union type parameters defined in the below given resource is not supported by ballerina HTTP module. 

```bal
import ballerina/http;

listener http:Listener ep0 = new (443, config = {host: "petstore3.swagger.io"});

service /api/v3 on ep0 {

    resource function get users/meetings/["Admin"|"HR"|"Engineering" group](
                                    ("available"|"pending")[] status, 
                                    @http:Header "UTC"|"LOCAL" 'x\-date\-format, 
                                    "scheduled"|"live"|"upcoming" 'type = "live") {
    }
}
```
![image](https://user-images.githubusercontent.com/34091327/233891315-e5794ba2-6c8c-48a6-a6af-4b7a04285a51.png)

Strangly, if these types defined as a TypeDefiniton and then reffered, the code is compiling successfully. 

Ex: 

```bal
public type TYPE "scheduled"|"live"|"upcoming";

service /api/v3 on ep0 {

    resource function get users/meetings/["Admin"|"HR"|"Engineering" group](TYPE 'type = "live") {
    }
}
```

Therefore, we are seeking for the support for following use cases in service resource functions. 

1. Union of singleton types support in header/query parameters
2. Array of union of singleton types support in header/query parameters - Ex: `("available"|"pending")[] status`
 

**Steps to reproduce:**

**Affected Versions:**

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

**Comment by TharmiganK** [comment]
This may be a bug in the compiler rewrite since we treat union types separately as we only allow unions with only nil type. Seems like we have missed to consider this union of singleton types case. Will fix that.

Btw `("available"|"pending")[] status` will not give any compiler errors with `2201.5.0`

And for the path parameter currently you will get a runtime error when starting the service like this:
```
error: incompatible path parameter type: '("Admin"|"HR"|"Engineering")'
```
This path param issue is tracked via https://github.com/ballerina-platform/ballerina-standard-library/issues/4372

**Comment by aneeshafedo** [comment]
@TharmiganK Would you be able to provide an ETA for the fix ?

**Comment by TharmiganK** [comment]
The following should be added:
1. runtime support for path parameter which is a sub type of one of `int`, `string`, `float`, `boolean` and `decimal`
2. casting query and header parameters to the type reference type of the parameter type. (Currently it only return the referred basic type corresponds to the parameter type)

Type conversion to the referenced type needed to be checked with lang. So the best case estimate for the above fixes will be Thursday (27th April, 2023).

Regarding the compiler validation errors for union type, there is a workaround to define a type with those union values and use that type in the resource signature. Also this workaround is more readable than having the type in-line. So we thought of holding that support until update 6 since we need some more effort to fix that. @aneeshafedo will that be ok?

**Comment by aneeshafedo** [comment]
> The following should be added:
> 
> 1. runtime support for path parameter which is a sub type of one of `int`, `string`, `float`, `boolean` and `decimal`
> 2. casting query and header parameters to the type reference type of the parameter type. (Currently it only return the referred basic type corresponds to the parameter type)
> 
> Type conversion to the referenced type needed to be checked with lang. So the best case estimate for the above fixes will be Thursday (27th April, 2023).
> 
> Regarding the compiler validation errors for union type, there is a workaround to define a type with those union values and use that type in the resource signature. Also this workaround is more readable than having the type in-line. So we thought of holding that support until update 6 since we need some more effort to fix that. @aneeshafedo will that be ok?

We considered the workaround and also mentioned that it is working in the issue itself. Here we are trying to map an enum given in the OpenAPI definition to Ballerina code. If the user has explicitly defined an enum we are currently generating the TypeDefinition and then refer to that. However, in the scenarios that user has not explicitly defined the enum instead defined inline the best way we can map to Ballerina is generating an inline type. Please refer the issue https://github.com/ballerina-platform/openapi-tools/issues/622#issuecomment-1512547954. 

Therefore workaround might not be the best option for the generated service which support a range of various enum scenarios.

**Comment by NipunaRanasinghe** [comment]
@aneeshafedo yeah as we discussed in https://github.com/ballerina-platform/openapi-tools/issues/622#issuecomment-1512547954, creating the named types by wrapping up the union type from the openAPI tool side may introduce a new set of issues (e.g. redeclared symbols). Also the workaround is less scalable in terms of maintainability (users will have to keep doing the same changes, whenever there's a newer contract version).

so IMO we better wait for the proper fixes from the HTTP side for the inline scenarios, as our proposed ETA is also update 6.

**Comment by aneeshafedo** [comment]
@NipunaRanasinghe yes, since our ETA is also update 6, if @TharmiganK can provide a timestamp of a build version once the fix is merged we can then continue our implementation in the planned approach and release our fix also with update 6. 

@TharmiganK is that possible and what would be the ETA to get the fix merged to Update 6 module.

**Comment by aneeshafedo** [comment]
> 

@TharmiganK could you please provide and example for the 2nd usecase here.

**Comment by TharmiganK** [comment]
@aneeshafedo Do you have use cases for these kind of query params in the resource functions:

```bal
type TYPE1 string;
type TYPE2 string;
type TYPE3 string?;

enum TYPE4 {
    VALUE1,
    VALUE2
}

type TYPE5 1|2|int;
```

**Case 1:** Union of subtypes
```bal
resource function get path1(TYPE1|TYPE2  val) {}

resource function get path2(TYPE1|TYPE2?  val) {}
```

**Case 2:** Union of nilable subtypes
```bal
resource function get path3(TYPE1|TYPE3  val) {}
```

**Case 3:** Union of `enum` and `string` types
```bal
resource function get path4(TYPE4|TYPE1 val) {}
```

**Case 4:** Union of finite types and matching basic types
```bal
resource function get path5(TYPE5 val) {}
```

**Comment by aneeshafedo** [comment]
@TharmiganK theoritacally speaking, possibe scenarios can be found for usecases 1, 2, and 4. Since we not generating enums currently usecase 3 is not possible. Anyway we do not support generation of any of these scenarios currently using the openapi tool when generating service from openapi definition.

**Comment by TharmiganK** [comment]
> @TharmiganK theoritacally speaking, possibe scenarios can be found in openapi definitions for usecases 1, 2, and 4. Since we not generating enums currently usecase 3 is not possible. Anyway we do not support generation of any of these scenarios currently using the openapi tool when generating service from openapi definition.

OK, actually these are valid types for a query parameter if we look at the effective type. But supporting them in the runtime will take some effort. Also supporting these types will affect the [constrained parameter support](https://github.com/ballerina-platform/ballerina-standard-library/issues/4371) we are planing for update 6. Hence, we will not allow these types for now and will return an appropriate compiler error. If there is a requirement, we will address in the later releases.

**Comment by TharmiganK** [comment]
@aneeshafedo This issue is fixed now. You can test it with the following timestamped version of `http` :
 `2.7.1-20230503-131100-9fadae2`

Since we have updated the dependent std lib timestamped versions in `http`, you may have to update open-api as well if you want to use this timestamped version. You can find the latest timestamped versions of other modules here: https://docs.google.com/spreadsheets/d/1kbqZ0ra4VxR2lVry73zt5nrNIcfgG5xmQ7VEepU0aog/edit#gid=2020497638
